lib/puppet/indirector/catalog/diff_compiler.rb

```diff
@@ -143,6 +143,12 @@ def add_node_data(node)
   # Compile baseline and preview catalogs
   #
   def compile(node, options)
+
+    if node.facts.nil? || node.facts.values.nil? || node.facts.values['osfamily'].nil?
+      # Node does not have a valid factset.
+      raise PuppetX::Puppetlabs::Preview::GeneralError, "Facts seems to be missing. No 'osfamily' fact found for node '#{name}'"
+    end
```

Comment:
Are you sure nodes have facts in all usage scenarios? As opposed to facts having been processed and set as node parameters and such (don't remember how /when exactly that happens, it may be the compiler that initiates that, and that this test needs to come later inside the preview derived compiler. (But I am not sure of the actual flow - what you  added may be right). Just saying if this is the cause of the test failures you were seeing...